docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+
+For example, the `system.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice` in the tree.
+
+A `slice` is nested in the hierarchy by it's naming convention.
+
+For example, the `system-foo.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice/system-foo.slice/`.
+
+Finally, to ensure everyone is on the same page, if `docker.service` is associated with `system.slice` and the user launches a container via
+`docker run -m 100M busybox`, a `scope` will be created because the process was not launched by `systemd` itself in the `slice` associated with the launching daemon.
+
+For example:
+
+`/sys/fs/cgroup/system.slice/docker-40c5264f71043fe15193603a05c7bf2a6b7f42b06951a7b59357cc7c2a380f04.scope`
+
+## kubelet cgroup driver
+
+The `kubelet` manipulates the `cgroup` file-system directly during initial bootstrapping.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="": driver to use for cgroup manipulation by the kubelet.  Possible values: cgroupfs, systemd`
+
+The `ContainerManager` interface in the `kubelet` should support an implementation for `cgroupfs` or `systemd`.
+
+Effort to share code across implementations should be made or separated out.
+
+For example, both implementations will most likely rely on a common set of kernel tunable flags.
+
+## kubelet cgroup bootstrapping (current behavior)
+
+During initial bootstrapping of the `kubelet`, it creates the following `cgroup` nodes:
+
+* `/kubelet`
+
+If the `docker` container runtime is used, it creates the following `cgroup` node, and attaches the `docker` pid.
+
+* `/docker-daemon` - a memory restriction is imposed that limits memory to max( machine capacity * .7, 150Mi), the OOM score is adjusted to -900
+
+If a `system-container` name is specified, it will create an additional `cgroup` node
+
+* `/<system-container-name>` - attaches all non-kernel host pids to facilitate accounting host daemon usage
+
+## kubelet cgroup bootstrapping under systemd
+
+If the `kubelet` is launched with the `--cgroup-driver=systemd`, the following behavior is proposed:
+
+* the `kubelet` detects the `.service` unit from which it was launched and uses that for monitoring its own resource usage
+
+If the `kubelet` is using the `docker` container runtime, the following behavior is proposed:
+
+* the `kubelet` detects the `.service` unit from which the `docker` daemon was launched (i.e. `docker.service`)
+
+If the `kubelet` wants to set memory limits on the `docker` daemon, it should only do this via a flag:
+
+* `--container-runtime-memory-threshold-limit="70": if specified, the kubelet will enforce memory limits on the container runtime agent`
+
+If specified, the `kubelet` should interface with the `systemd` API to set runtime scoped limits on the `docker.service` directly.  If not
+specified, the `kubelet` will delegate to whatever value was set directly via the administrator that configured the `docker.service`.
```

Comment:
nit, it's not really delegating, it's just not doing anything if `docker.service` happened to have something configured.